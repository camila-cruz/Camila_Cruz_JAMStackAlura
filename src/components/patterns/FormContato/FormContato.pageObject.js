export default class FormContatoPageObject {
  constructor(cy) {
    this.cy = cy;

    this.cy.visit('/');
  }

  acessaForm() {
    this.cy.get('#btnContato').click();
    return this;
  }

  preencheForm({ name, email, message }) {
    this.cy.get('input[name="name"]').type(name);
    this.cy.get('input[name="email"]').type(email);
    this.cy.get('textarea[name="message"]').type(message);

    return this;
  }

  enviaForm() {
    this.cy.get('button[type="submit"]').click();
    return this;
  }
}
