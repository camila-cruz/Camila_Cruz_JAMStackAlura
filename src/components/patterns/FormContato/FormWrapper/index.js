/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Grid } from '../../../foundation/layout/Grid';

const FormBase = styled.div`
  position: relative;
  display: flex;
  flex: 1;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  align-self: center;

  padding-top: 80px;
  padding-bottom: 80px;

  background-color: ${({ theme }) => theme.background.primary};
  color: ${({ theme }) => theme.text.primary};

  max-width: 600px;
  /* 
  height: 100%;
  clip-path: polygon(25% 10%, 75% 10%, 100% 50%, 75% 90%, 25% 90%, 0% 50%); */

  & > form {
    min-width: 350px;
    width: auto;
    max-width: 400px;
    padding-left: 20px;
    padding-right: 20px;
}
`;

export default function FormWrapper({ children, ...props }) {
  return (
    <Grid.Row
      marginLeft={0}
      marginRight={0}
      flex={1}
      justifyContent="center"
    >
      <Grid.Col
        value={{ xs: 12, md: 7, lg: 5 }}
        alignSelf="center"
      >
        <FormBase {...props}>
          {children}
        </FormBase>
      </Grid.Col>
    </Grid.Row>
  );
}

FormWrapper.propTypes = {
  children: PropTypes.node.isRequired,
};
