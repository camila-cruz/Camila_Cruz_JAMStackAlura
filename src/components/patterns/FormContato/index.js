/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState } from 'react';
import styled from 'styled-components';
import { Grid } from '../../foundation/layout/Grid';
import { Button } from '../../commons/Button';
import TextField from '../../forms/TextField';

const FormWrapper = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  align-self: center;

  padding-top: 80px;
  padding-bottom: 80px;

  /* 
  height: 100%;
  clip-path: polygon(25% 10%, 75% 10%, 100% 50%, 75% 90%, 25% 90%, 0% 50%); */
  background-color: white;
`;

function FormContent() {
  const [messageInfo, setMessageInfo] = useState({
    name: '',
    email: '',
    message: '',
  });

  const messageDTO = {
    name: messageInfo.name,
    email: messageInfo.email,
    message: messageInfo.email,
  };

  function handleChange(e) {
    const fieldName = e.target.getAttribute('name');

    setMessageInfo({
      ...messageInfo,
      [fieldName]: e.target.value,
    });
  }

  function onSubmit(e) {
    e.preventDefault();

    fetch('https://contact-form-api-jamstack.herokuapp.com/message', {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(messageDTO),
    })
      .then((res) => res.json())
      .then((data) => {
        // eslint-disable-next-line no-console
        console.log(data);
      });
  }

  return (
    <form onSubmit={onSubmit}>
      <h2>Envie sua mensagem</h2>

      <div>
        <label htmlFor="name">
          Nome
          <TextField type="text" name="name" id="name" value={messageInfo.name} onChange={handleChange} />
        </label>
      </div>

      <div>
        <label htmlFor="email">
          E-mail
          <TextField type="email" name="email" id="email" value={messageInfo.email} onChange={handleChange} />
        </label>
      </div>

      <div>
        <label htmlFor="message">
          Mensagem
          <TextField type="text" name="message" id="message" value={messageInfo.message} onChange={handleChange} textarea />
        </label>
      </div>

      <Button type="submit" variant="tertiary" fullWidth>
        Enviar
      </Button>
    </form>
  );
}

// eslint-disable-next-line react/prop-types
export default function FormContato({ props }) {
  return (
    <Grid.Row
      marginLeft={0}
      marginRight={0}
      flex={1}
      justifyContent="center"
    >
      <Grid.Col
        value={{ xs: 12, md: 6, lg: 4 }}
        alignSelf="center"
      >
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        <FormWrapper {...props}>
          <FormContent />
        </FormWrapper>
      </Grid.Col>
    </Grid.Row>
  );
}
