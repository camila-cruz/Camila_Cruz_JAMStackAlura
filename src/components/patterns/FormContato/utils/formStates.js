import loadingAnimation from '../animations/loading.json';
import successAnimation from '../animations/success.json';
import errorAnimation from '../animations/error.json';

export default {
  DEFAULT: {
    feedback: {
      animation: '',
      loop: false,
    },
    buttonText: 'Enviar mensagem',
  },
  LOADING: {
    feedback: {
      animation: loadingAnimation,
      loop: true,
    },
    text: 'Aguarde',
  },
  DONE: {
    feedback: {
      animation: successAnimation,
      loop: false,
    },
    text: 'Mensagem enviada!',
    buttonText: 'Nova mensagem',
  },
  ERROR: {
    feedback: {
      animation: errorAnimation,
      loop: false,
    },
    text: 'Houve um erro!',
    buttonText: 'Tente novamente',
  },
};
