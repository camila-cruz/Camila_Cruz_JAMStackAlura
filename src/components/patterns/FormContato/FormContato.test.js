import React from 'react';
import user from '@testing-library/user-event';
import {
  render, screen, act, waitFor,
} from '../../../infra/test/testUtils';
import FormContato from '.';
import formStates from './utils/formStates';

const nameMock = 'Astrogilda';
const emailMock = 'astro@gilda.com';
const messageMock = 'OlÃ¡, mundo!';
const onSubmitMock = jest.fn((e) => e.preventDefault());

describe('<FormContato />', () => {
  describe('when the user fills the form', () => {
    describe('with valid data', () => {
      test('it attempts to submit', async () => {
        await act(async () => render(
          <FormContato
            CloseButton={() => null}
            props={{}}
            onSubmit={onSubmitMock}
          />,
        ));

        const submitButton = screen.getByRole('button', {
          name: /enviar/i,
        });
        expect(submitButton).toBeDisabled();

        const nameInput = screen.getByLabelText(/nome/i);
        user.type(nameInput, nameMock);
        await waitFor(() => expect(nameInput).toHaveValue(nameMock));

        const emailInput = screen.getByLabelText(/e-mail/i);
        user.type(emailInput, emailMock);
        await waitFor(() => expect(emailInput).toHaveValue(emailMock));

        const messageInput = screen.getByLabelText(/mensagem/i);
        user.type(messageInput, messageMock);
        await waitFor(() => expect(messageInput).toHaveValue(messageMock));

        expect(submitButton).not.toBeDisabled();

        user.click(submitButton);
        expect(onSubmitMock).toHaveBeenCalledTimes(1);
      });

      test('a loading animation is displayed', async () => {
        await act(async () => render(
          <FormContato
            CloseButton={() => null}
            props={{}}
            onSubmit={onSubmitMock}
            formSubmission
            formState={formStates.LOADING}
          />,
        ));

        const alertMessage = screen.getByRole('alert');
        expect(alertMessage).toMatchSnapshot();
      });

      describe('when the submission completes', () => {
        test('a feedback message is displayed', async () => {
          await act(async () => render(
            <FormContato
              CloseButton={() => null}
              props={{}}
              onSubmit={onSubmitMock}
              formSubmission
              formState={formStates.DONE}
            />,
          ));

          const alertMessage = screen.getByRole('alert');
          expect(alertMessage).toMatchSnapshot();

          const actionButton = screen.getByRole('button');
          await act(async () => user.click(actionButton));

          expect(screen.getByRole('button', {
            name: /enviar/i,
          })).toBeDefined();
        });
      });
    });

    describe('with invalid data', () => {
      test('the errors are displayed', async () => {
        await act(async () => render(
          <FormContato
            CloseButton={() => null}
            props={{}}
            onSubmit={onSubmitMock}
          />,
        ));

        const emailInput = screen.getByLabelText(/e-mail/i);
        emailInput.focus();
        emailInput.blur();

        const errorAlert = screen.getByRole('alert');
        expect(errorAlert).toBeDefined();
      });
    });
  });
});
