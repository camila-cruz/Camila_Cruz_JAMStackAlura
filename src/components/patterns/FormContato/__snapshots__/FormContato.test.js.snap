// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormContato /> when the user fills the form with valid data a loading animation is displayed 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 300px;
}

.c1 {
  font-size: 32px;
  line-height: normal;
}

@media screen and (min-width:0px) {

}

@media screen and (min-width:768px) {

}

@media screen and (min-width:992px) {

}

<div
  class="c0"
  display="flex"
  role="alert"
>
  <div
    class="lottie-container basic"
    style="width: 150px; height: 150px;"
  >
    <svg
      height="172"
      preserveAspectRatio="xMidYMid meet"
      style="width: 100%; height: 100%; transform: translate3d(0,0,0);"
      viewBox="0 0 172 172"
      width="172"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <clippath
          id="__lottie_element_2"
        >
          <rect
            height="172"
            width="172"
            x="0"
            y="0"
          />
        </clippath>
      </defs>
      <g
        clip-path="url(#__lottie_element_2)"
      >
        <g
          opacity="1"
          style="display: block;"
          transform="matrix(0.26875001192092896,0,0,0.26875001192092896,86,86)"
        >
          <g
            opacity="1"
            transform="matrix(1,0,0,1,0,0)"
          >
            <path
              d=" M0,-180 C3.7960000038146973,-180 7.565000057220459,-179.8820037841797 11.303000450134277,-179.64999389648438"
              fill-opacity="0"
              stroke="rgb(244,156,16)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-opacity="1"
              stroke-width="36"
            />
          </g>
        </g>
        <g
          opacity="1"
          style="display: block;"
          transform="matrix(0.26875001192092896,0,0,0.26875001192092896,86,86)"
        >
          <g
            opacity="1"
            transform="matrix(-1,0,0,-1,0,0)"
          >
            <path
              d=" M0,-100 C2.1089999675750732,-100 4.203000068664551,-99.93499755859375 6.28000020980835,-99.80599975585938"
              fill-opacity="0"
              stroke="rgb(219,88,11)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-opacity="1"
              stroke-width="36"
            />
          </g>
        </g>
      </g>
    </svg>
  </div>
  <h2
    class="c1"
    size="1"
  >
    Aguarde
  </h2>
</div>
`;

exports[`<FormContato /> when the user fills the form with valid data when the submission completes a feedback message is displayed 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 300px;
}

.c2 {
  width: 200px;
  height: 48px;
  font-size: 16px;
  font-weight: 900;
  text-transform: uppercase;
  cursor: pointer;
  border: 0;
  border-radius: 8px;
  background-color: #512919;
  color: #F3CF77;
  margin-top: 20px;
  -webkit-transition: 0.3s ease;
  transition: 0.3s ease;
}

.c2:hover {
  background-color: #fff4b2;
  color: #512919;
}

.c1 {
  font-size: 32px;
  line-height: normal;
}

@media screen and (min-width:0px) {

}

@media screen and (min-width:768px) {

}

@media screen and (min-width:992px) {

}

<div
  class="c0"
  display="flex"
  role="alert"
>
  <div
    class="lottie-container basic"
    style="width: 150px; height: 150px;"
  >
    <svg
      height="200"
      preserveAspectRatio="xMidYMid meet"
      style="width: 100%; height: 100%; transform: translate3d(0,0,0);"
      viewBox="0 0 200 200"
      width="200"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <clippath
          id="__lottie_element_10"
        >
          <rect
            height="200"
            width="200"
            x="0"
            y="0"
          />
        </clippath>
      </defs>
      <g
        clip-path="url(#__lottie_element_10)"
      >
        <g
          style="display: none;"
        >
          <g>
            <path />
          </g>
        </g>
        <g
          opacity="1"
          style="display: block;"
          transform="matrix(1.0050300359725952,0,0,1.0050300359725952,102.75800323486328,103.76300048828125)"
        >
          <g
            opacity="1"
            transform="matrix(0.9776999950408936,0,0,0.9776999950408936,-2.74399995803833,-3.74399995803833)"
          >
            <path
              d="M0 0"
              fill-opacity="0"
              stroke="rgb(96,191,131)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-opacity="1"
              stroke-width="6"
            />
          </g>
        </g>
        <g
          opacity="1"
          style="display: block;"
          transform="matrix(0.7477700114250183,0,0,0.7477700114250183,100,99.625)"
        >
          <g
            opacity="1"
            transform="matrix(0.9924899935722351,0,0,0.9763200283050537,0.3330000042915344,1)"
          >
            <path
              d="M0 0"
              fill-opacity="0"
              stroke="rgb(96,191,131)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-opacity="1"
              stroke-width="8"
            />
          </g>
        </g>
        <g
          opacity="1"
          style="display: block;"
          transform="matrix(0.7477700114250183,0,0,0.7477700114250183,100,99.625)"
        >
          <g
            opacity="1"
            transform="matrix(0.9924899935722351,0,0,0.9763200283050537,0.3330000042915344,1)"
          >
            <path
              d="M0 0"
              fill-opacity="0"
              stroke="rgb(96,191,131)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-opacity="1"
              stroke-width="8"
            />
          </g>
          <g
            opacity="1"
            transform="matrix(1,0,0,1,0,0)"
          >
            <path
              d=" M-189.21600341796875,26.12700080871582"
              fill-opacity="0"
              stroke="rgb(96,191,131)"
              stroke-linecap="butt"
              stroke-linejoin="miter"
              stroke-miterlimit="4"
              stroke-opacity="1"
              stroke-width="8"
            />
          </g>
        </g>
      </g>
    </svg>
  </div>
  <h2
    class="c1"
    size="1"
  >
    Mensagem enviada!
  </h2>
  <button
    class="c2"
  >
    Nova mensagem
  </button>
</div>
`;
